pkg/registration/hub/importer/importer.go

```diff
@@ -0,0 +1,284 @@
+	"open-cluster-management.io/ocm/pkg/operator/helpers/chart"
+	cloudproviders "open-cluster-management.io/ocm/pkg/registration/hub/importer/providers"
+)
+
+const (
+	operatorNamesapce     = "open-cluster-management"
+	bootstrapSA           = "cluster-bootstrap"
+	ConditionTypeImported = "Imported"
+)
+
+var (
+	genericScheme = runtime.NewScheme()
+	genericCodecs = serializer.NewCodecFactory(genericScheme)
+	genericCodec  = genericCodecs.UniversalDeserializer()
+)
+
+func init() {
+	utilruntime.Must(api.InstallKube(genericScheme))
+	utilruntime.Must(apiextensionsv1.AddToScheme(genericScheme))
+	utilruntime.Must(operatorv1.Install(genericScheme))
+}
+
+// KlusterletConfigRenderer renders the config for klusterlet chart.
+type KlusterletConfigRenderer func(
+	ctx context.Context, config *chart.KlusterletChartConfig) (*chart.KlusterletChartConfig, error)
+
+type Importer struct {
+	providers     []cloudproviders.Interface
+	clusterClient clusterclientset.Interface
+	clusterLister clusterlisterv1.ManagedClusterLister
+	renders       []KlusterletConfigRenderer
+	patcher       patcher.Patcher[*v1.ManagedCluster, v1.ManagedClusterSpec, v1.ManagedClusterStatus]
+}
+
+// NewImporter creates an auto import controller
+func NewImporter(
+	renders []KlusterletConfigRenderer,
+	clusterClient clusterclientset.Interface,
+	clusterInformer clusterinformerv1.ManagedClusterInformer,
+	providers []cloudproviders.Interface,
+	recorder events.Recorder) factory.Controller {
+	controllerName := "managed-cluster-importer"
+	syncCtx := factory.NewSyncContext(controllerName, recorder)
+
+	i := &Importer{
+		providers:     providers,
+		clusterClient: clusterClient,
+		clusterLister: clusterInformer.Lister(),
+		renders:       renders,
+		patcher: patcher.NewPatcher[
+			*v1.ManagedCluster, v1.ManagedClusterSpec, v1.ManagedClusterStatus](
+			clusterClient.ClusterV1().ManagedClusters()),
+	}
+
+	for _, provider := range providers {
+		provider.Register(syncCtx)
+	}
+
+	return factory.New().WithInformersQueueKeysFunc(queue.QueueKeyByMetaName, clusterInformer.Informer()).
+		WithSyncContext(syncCtx).WithSync(i.sync).ToController(controllerName, recorder)
+}
+
+func (i *Importer) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	clusterName := syncCtx.QueueKey()
+	logger := klog.FromContext(ctx)
+	logger.V(4).Info("Reconciling key", "clusterName", clusterName)
+
+	cluster, err := i.clusterLister.Get(clusterName)
+	switch {
+	case errors.IsNotFound(err):
+		return nil
+	case err != nil:
+		return err
+	}
+
+	// If the cluster is imported, skip the reconcile
+	if meta.IsStatusConditionTrue(cluster.Status.Conditions, ConditionTypeImported) {
+		return nil
+	}
+
+	// get provider from the provider list
+	var provider cloudproviders.Interface
+	for _, p := range i.providers {
+		if p.IsManagedClusterOwner(cluster) {
+			provider = p
+		}
+	}
+	if provider == nil {
+		logger.V(4).Info("provider not found for cluster", "cluster", cluster.Name)
```

Comment:
```suggestion
		logger.Info("provider not found for cluster", "cluster", cluster.Name)
```

Reply:
maybe V(2), otherwise when we enable the controller, this logs will always be printed since it watches managedCluster